### PR TITLE
refactor: rewrite revert and precompile decoding

### LINKED ISSUE #5337: ecpairing precompile decoding is incorrect
### Component

Forge, Other (please describe)

### Have you ensured that all of these are up to date?

- [ ] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (c78a811 2023-07-08T00:13:27.174343000Z)

### What command(s) is the bug in?

forge test

### Operating System

macOS (Apple Silicon)

### Describe the bug

https://github.com/foundry-rs/foundry/blob/c78a811a8af95fb1e029427583a07b2ca3a3fa51/evm/src/trace/decoder.rs#L130

I'm not sure where `x1, y1, x3, x2, y3, y2` comes from, but it's there.

The ecpairing interface doesn't follow the ABI standard strictly and is `((x, y), ((x0, x1), (y0, y1)))*`. See [eip-197].

[eip-197]: https://eips.ethereum.org/EIPS/eip-197#encoding

### PR DESCRIPTION
**Pull request by DaniPopes**
## Motivation

Closes https://github.com/foundry-rs/foundry/issues/6102
Fixes #5337

### CODE REVIEW

File: crates/evm/evm/src/executors/invariant/mod.rs [hunk 1]
```diff
@@ -29,7 +29,7 @@ use proptest::{
     test_runner::{TestCaseError, TestRunner},
 };
 use revm::{
-    primitives::{Address as rAddress, HashMap, U256 as rU256},
+    primitives::{Address as aAddress, HashMap, U256 as rU256},
```
Comment (Evalir) [inline crates/evm/evm/src/executors/invariant/mod.rs:32]: I think we can just use `alloy_primitives::Address` now

File: crates/evm/traces/src/decoder/precompiles.rs [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+            let (sig, ecmulCall { x1, y1, s }) = tri!(abi_decode_call(data));
+            (sig, vec![x1.to_string(), y1.to_string(), s.to_string()])
+        }
+        0x08 => (ecpairingCall::SIGNATURE, tri!(decode_ecpairing(data))),
+        0x09 => (blake2fCall::SIGNATURE, tri!(decode_blake2f(data))),
+        0x0a => (pointEvaluationCall::SIGNATURE, tri!(decode_kzg(data))),
+        _ => unreachable!(),
+    };
+
+    // TODO: Other chain precompiles
```
Comment (Evalir) [inline crates/evm/traces/src/decoder/precompiles.rs:68]: I'm a bit unsure if we want to implement precompiles for other chains—we had discussed this is a thing we probably don't want to do (albeit we could probably do Optimism's easily?) cc @mattsse

File: crates/evm/traces/src/decoder/precompiles.rs [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+        0x05 => (modexpCall::SIGNATURE, tri!(decode_modexp(data))),
+        0x06 => {
+            let (sig, ecaddCall { x1, y1, x2, y2 }) = tri!(abi_decode_call(data));
+            (sig, vec![x1.to_string(), y1.to_string(), x2.to_string(), y2.to_string()])
+        }
+        0x07 => {
+            let (sig, ecmulCall { x1, y1, s }) = tri!(abi_decode_call(data));
+            (sig, vec![x1.to_string(), y1.to_string(), s.to_string()])
+        }
+        0x08 => (ecpairingCall::SIGNATURE, tri!(decode_ecpairing(data))),
```
Comment (Evalir) [inline crates/evm/traces/src/decoder/precompiles.rs:62]: I think this also solves #5337

File: crates/evm/traces/src/decoder/mod.rs [hunk 4]
```diff
@@ -241,28 +206,37 @@ impl CallTraceDecoder {
...
 
+            // Decode events
+            self.decode_events(node).await;
+
             // Decode call
-            if let Some(precompile_fn) = self.precompiles.get(&node.trace.address) {
-                node.decode_precompile(precompile_fn, &self.labels);
-            } else if let RawOrDecodedCall::Raw(ref bytes) = node.trace.data {
-                if bytes.len() >= 4 {
+            // TODO: chain ID argument
```
Comment (Evalir) [inline crates/evm/traces/src/decoder/mod.rs:226]: This is to decode different precompiles per chain id?

**Review by Evalir — COMMENTED** [review]
Super cool. Some comments:

File: crates/evm/traces/src/decoder/mod.rs (hunk 4 above)
Comment (DaniPopes) [reply crates/evm/traces/src/decoder/mod.rs:226]: Yes that's what I was thinking, but maybe I'm thinking ahead too much. Maybe we should just try to decode whatever revm has set as precompiles

File: crates/evm/evm/src/executors/invariant/mod.rs (hunk 1 above)
Comment (DaniPopes) [reply crates/evm/evm/src/executors/invariant/mod.rs]: Anvil stuff, I don't want to get into that

File: crates/evm/traces/src/decoder/precompiles.rs (hunk 2 above)
Comment (DaniPopes) [reply crates/evm/traces/src/decoder/precompiles.rs:68]: https://github.com/foundry-rs/foundry/issues/1078

File: crates/evm/traces/src/decoder/mod.rs (hunk 4 above)
Comment (Evalir) [reply crates/evm/traces/src/decoder/mod.rs:226]: Yeah I think might be best to stick to what `revm` has—I think soft consensus is to not get into maintaining precompiles for other chains—they can become a pain. IIRC arbitrum, which is the biggest offender, has a repo that simulates their custom precompile that ppl can use

Comment (DaniPopes) [reply crates/evm/traces/src/decoder/mod.rs:226]: Ref https://github.com/foundry-rs/foundry/issues/748

**Review by Evalir — APPROVED** [review]
i'm cool with the changes, the only contentious thing is decoding other precompiles—but as this is not implemented imho not a blocker

**Comment by DaniPopes** [comment]
A thing to note is that now `Error(string)`/`Panic(uint256)` get displayed as `revert: <msg>` and `panic: <msg>` rather than just `<msg>`.

**Review by mattsse — APPROVED** [review]

**Pull request merged**